### PR TITLE
[Core] Add soft thinking token budget with progressive logit bias

### PR DESCRIPTION
**Pull request by efortin**
## Summary

This PR keeps `thinking_token_budget` as a hard safety wall, but adds a soft path for standard decoding before that wall is reached.

- In the last 30% of the budget, `ThinkingBudgetStateHolder` updates `soft_progress` every token instead of letting the countdown early-return skip the soft zone.
- The logits path nudges only the next valid end-of-thinking token, so multi-token close markers are completed in order.
- The nudge stays in torch tensor ops, with no `.item()` sync in the sampling loop.
- Speculative decoding keeps the existing hard-force behavior; the soft bias is skipped there.
- At 100% of the budget, the existing hard force remains the safety net.

## Why

A pure hard cut can land mid-thought. When that happens, the model may not realize its thinking section was interrupted and can leak reasoning into the visible answer/content. The soft zone gives it a chance to close naturally before the hard budget cut.

## Duplicate Check

I checked open PRs for `thinking token budget` and `soft thinking budget`. I did not find another open PR adding this soft close-token bias to `ThinkingBudgetStateHolder`; the closest open PRs cover MRV2 support, enforcement bugs, implicit tool calls, parser detection, or speculative decoding acceptance.

## Validation

- `UV_CACHE_DIR=/private/tmp/uv-cache PYTHONPATH=/private/tmp/vllm-test-sitecustomize:/Users/manu/Projects/vllm VLLM_TARGET_DEVICE=cpu uv run --active --no-sync python -m pytest tests/v1/logits_processors/test_correctness.py -q -k "thinking_budget"` -> 15 passed, 15 deselected
- `UV_CACHE_DIR=/private/tmp/uv-cache uv run --active --no-sync python -m compileall -q vllm/v1/sample/thinking_budget_state.py tests/v1/logits_processors/test_correctness.py`
- `git diff --check`

The local test run used a temporary `sitecustomize` only to force vLLM's CPU platform on macOS arm64, because no editable/precompiled vLLM wheel is available for this local platform.

## References

- NVIDIA NIM BudgetControlLogitsProcessor: https://docs.nvidia.com/nim/large-language-models/1.13.0/thinking-budget-control.html
- BudgetThinker: https://arxiv.org/html/2508.17196v1

AI assistance was used; the human author should review the changed lines and validation before reopening or requesting review.

### CODE REVIEW

File: vllm/v1/sample/logits_processor/builtin.py [hunk 1]
```diff
@@ -457,15 +459,29 @@ def _update_think_state(self, state: dict[str, Any]):
...
-                state["check_count_down"] = state["thinking_token_budget"]
+                state["check_count_down"] = budget
+            elif state["in_think"] and count >= soft_start and budget > 0:
+                # Soft zone: ramp bias from 0 to 8 over last 20%
+                progress = (count - soft_start) / max(1, budget - soft_start)
+                state["in_soft"] = True
+                state["soft_bias"] = progress * 8.0
+            else:
+                state["in_soft"] = False
+                state["soft_bias"] = 0.0
```
[the remark this answers is not in the document]
Comment (efortin) [reply vllm/v1/sample/logits_processor/builtin.py]: When budget is enabled, the update_state is called each token.

**Comment by efortin** [comment]
@claude review

**Comment by efortin** [comment]
Rebased onto current main.

Since this PR was opened, #34668 moved the thinking-budget logic from `logits_processor/builtin.py` into `ThinkingBudgetStateHolder` (`vllm/v1/sample/thinking_budget_state.py`) with speculative-decoding support, which is why the branch was conflicting. The soft zone is now implemented against that holder:

- `soft_progress` state ramps 0→1 over the last 30% of the budget in `_update_think_state`, with resets on every exit path (natural close, hard-force transition, end-sequence completion).
- The countdown early-return stays active only while the remaining budget is outside the soft window, so the ramp updates every step inside the zone.
- The adaptive bias (`target = end_logit + 2 * gap * progress`) is applied in `_apply_forcing_to_logits` on the standard decoding path; with speculative decoding enabled the soft bias is skipped and the hard force from #34668 applies unchanged.

Also validated the same formula independently on an MLX-based OpenAI-compatible server (jundot/omlx#1812) with Qwen3.6-35B: 14/14 structured-output calls across budgets 50–600, zero content leaks, thinking length scaling smoothly with the budget, and natural `</think>` closes inside the soft zone.
